Add note about sysimage `__init__`s running before startup.jl (#51623)
JuliaLang/julia: doc/src/manual/environment-variables.md

!!! note

    It is recommended to avoid changing environment variables during runtime,
    such as within a `~/.julia/config/startup.jl`.

    One reason is that some julia language variables, such as [`JULIA_NUM_THREADS`](@ref JULIA_NUM_THREADS)
    and [`JULIA_PROJECT`](@ref JULIA_PROJECT), need to be set before Julia starts.

    Similarly, `__init__()` functions of modules in the sysimage are run before `startup.jl`,

>     so setting environment variables in a `startup.jl` may be too late for user code.

Reviewer: I am still not a fan of the "in the sysimage", I think the more common scenario is the `init` functions of modules being used from startup.jl.

Author: A user can see the order of things in a startup.jl but they may not be aware that sysimage module units are run before startup.jl is run. 

I wanted to add this specifically as it caught me out.

Author: How about 
```suggestion
    Similarly, `__init__()` functions of user modules in the sysimage (via PackageCompiler) are 
    run before `startup.jl`, so setting environment variables in a `startup.jl` may be too late for 
    user code.
```

!!! note

    Some variables, such as [`JULIA_NUM_THREADS`](@ref JULIA_NUM_THREADS) and
    [`JULIA_PROJECT`](@ref JULIA_PROJECT), need to be set before Julia
    starts. Adding these to `~/.julia/config/startup.jl` is too late in the startup process.

>     Also `__init__()` functions of modules in the sysimage will not see environment variables
>     that are set in `startup.jl` as they are run beforehand.

Reviewer: ```suggestion
    Also `__init__()` functions of modules may not see environment variables
    that are set in `startup.jl` as they are may be run beforehand.
```

E.g. if your startup.jl file does a `using`.

Author: I think the user can figure out if they're setting an ENV entry after a `using` in startup.jl. This is more about the sysimage gotcha

Reviewer: Eh, generally speaking depending on the order seems iffy, and the sysimg has little to do with that.

Author: Maybe the advice should be generalized to avoid setting ENV in startup.jl, motivated by things like sysimage inits running before startup.jl

Reviewer 2: It can also be worth noting that changing an environment variable at all can introduce data races into otherwise benign code, so is best carefully avoided as much as possible.

>     Some variables, such as [`JULIA_NUM_THREADS`](@ref JULIA_NUM_THREADS) and
>     [`JULIA_PROJECT`](@ref JULIA_PROJECT), need to be set before Julia
>     starts. Adding these to `~/.julia/config/startup.jl` is too late in the startup process.
> 
>     Also `__init__()` functions of modules in the sysimage will not see environment variables
>     that are set in `startup.jl` as they are run beforehand.

Author: ```suggestion
    It is recommended to avoid changing environment variables during runtime,
    such as within a `~/.julia/config/startup.jl`.

    One reason is that some julia language variables, such as [`JULIA_NUM_THREADS`](@ref JULIA_NUM_THREADS)
    and [`JULIA_PROJECT`](@ref JULIA_PROJECT), need to be set before Julia starts.

    Similarly, `__init__()` functions of modules in the sysimage are run before `startup.jl`,
    so setting environment variables in a `startup.jl` may be too late for user code.

    Further, changing environment variables during runtime can introduce data races into
    otherwise benign code.
```
